mcs/class/System/System.Net/WebRequestStream.cs

```
@@ -0,0 +1,436 @@
+			if (sendChunked || !HasWriteBuffer)
+				return;
+
+			BufferOffsetSize buffer = GetWriteBuffer ();
+			if (buffer != null && !Operation.IsNtlmChallenge && Request.ContentLength != -1 && Request.ContentLength < buffer.Size) {
+				closed = true;
+				var throwMe = new WebException ("Specified Content-Length is less than the number of bytes to write", null,
+					WebExceptionStatus.ServerProtocolViolation, null);
+				Operation.CompleteRequestWritten (this, throwMe);
+				throw throwMe;
+			}
+
+			await SetHeadersAsync (true, cancellationToken).ConfigureAwait (false);
+
+			WebConnection.Debug ($"{ME} WRITE REQUEST #1: {buffer != null}");
+
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+			if (buffer != null && buffer.Size > 0)
+				await InnerStream.WriteAsync (buffer.Buffer, 0, buffer.Size, cancellationToken);
+
+			await FinishWriting (cancellationToken);
+		}
+
+		async Task WriteChunkTrailer_inner (CancellationToken cancellationToken)
+		{
+			if (Interlocked.CompareExchange (ref chunkTrailerWritten, 1, 0) != 0)
+				return;
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+			byte[] chunk = Encoding.ASCII.GetBytes ("0\r\n\r\n");
+			await InnerStream.WriteAsync (chunk, 0, chunk.Length, cancellationToken).ConfigureAwait (false);
+		}
+
+		async Task WriteChunkTrailer ()
+		{
+			using (var cts = new CancellationTokenSource ()) {
+				cts.CancelAfter (WriteTimeout);
+				var timeoutTask = Task.Delay (WriteTimeout);
+				while (true) {
+					var myWriteTcs = new TaskCompletionSource<int> ();
+					var oldTcs = Interlocked.CompareExchange (ref pendingWrite, myWriteTcs, null);
+					if (oldTcs == null)
+						break;
+					var ret = await Task.WhenAny (timeoutTask, oldTcs.Task).ConfigureAwait (false);
+					if (ret == timeoutTask)
+						throw new WebException ("The operation has timed out.", WebExceptionStatus.Timeout);
+				}
+
+				try {
+					await WriteChunkTrailer_inner (cts.Token).ConfigureAwait (false);
+				} catch {
+					;
```

Why are we hiding exceptions here?

---

Intentional, going to stay.  I'll mark it in code.  This is very broken anyways.

---

See comment (and FIXME) in line 408.